### ISSUE TITLE
Locale formatting on Windows Server

### ISSUE THREAD

**Issue by danikuci1** [firefly-iii/firefly-iii#4012]
I am running Firefly III version 5.4.4

**Description**
when loading any page in firefly iii, from any browser, a banner appears which states `Invalid server configuration:  Firefly III is unable to format monetary amounts because your server is missing the required packages. There are instructions how to do this.`

I've read the instructions in the link, but they are for ubuntu, and I am running firefly iii in a WAMP stack on Windows Server. It's a US english install of Windows, and that's the language and locale I want to use in Firefly III. It seems that maybe Firefly III is unable to check this info properly on a Windows machine, but I am not sure where it is doing so.

**Extra info**
Debug information generated at 2020-10-29 08:51:19 America/Los_Angeles for Firefly III version **5.4.4**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.4.4 |
| Firefly III API | 1.4.0 |
| PHP | 7.4.11 |
| Host | WINNT |

| System info | Value |
| --- | --- |
| Installation ID | f43bb6ce-75a0-4995-bcbd-190833069c33 |
| Using docker? | false |
| Telemetry | false |
| Layout | v1 |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | en_US |
| BCscale | 12 |
| DB drivers | mysql |
| Current driver | mysql |
| DB version | 15 (exp. 15) |
| Login provider | eloquent  |
| Trusted proxies (.env) | 10.0.0.201 |

| User info | Value |
| --- | --- |
| Session start | 2020-10-01 00:00:00 |
| Session end | 2020-10-31 23:59:59 |
| Session first | 2020-10-29 08:34:26 |
| User ID | 1 |
| User language | en_US |
| User locale | en-US |
| Attempt at "en-US.utf8" | 'en-US.utf8' |
| Attempt at "en-US.UTF-8" | 'en-US.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36 |

-----
the following error repeats numerous times in the log:

[2020-10-29 08:34:00] local.ERROR: Could not set locale. The following array doesnt work:  ["en-US.utf8","en-US.UTF-8"] 

My question is: where or how can I adjust something so that firefly iii is aware of what the system locale is?

**Comment by sephrat** [comment]
Not sure what could be wrong with your config, but mine works fine on Windows 10 in fr-FR/en-US/en-GB after I installed the languages in Windows settings:
![image](https://user-images.githubusercontent.com/34862846/97611466-82bd9f80-1a16-11eb-927a-99946b5890b7.png)


| Scope | Version |
| --- | --- |
| Firefly III | 5.4.5 |
| Firefly III API | 1.4.0 |
| PHP | 7.4.6 |
| Host | WINNT |

| User info | Value |
| --- | --- |
| Session start | 2020-10-01 00:00:00 |
| Session end | 2020-10-31 23:59:59 |
| Session first | 2020-08-02 15:45:24 |
| User ID | 1 |
| User language | fr_FR |
| User locale | fr-FR |
| Attempt at "fr-FR.utf8" | 'fr-FR.utf8' |
| Attempt at "fr-FR.UTF-8" | 'fr-FR.UTF-8' |
```
[2020-10-29 18:35:01] local.DEBUG: Trying to set fr-FR.utf8  
[2020-10-29 18:35:01] local.DEBUG: Trying to set fr-FR.UTF-8  
```

**Comment by danikuci1** [comment]
my install is using default language settings from windows media, which is set to US english:

![image](https://user-images.githubusercontent.com/5247319/97627361-84706e80-19e8-11eb-9aa3-1b05cec0a6a3.png)

**Comment by sephrat** [comment]
Are you a new user of Firefly III? Or did it use to work on your machine before an upgrade? 
I'm asking this because in #3638 we made it so that locales on Windows are "en-US" instead of "en_US". Maybe on some setup "en_US" is the correct choice for whatever reason. 

What version of Windows are you running?
You could try enabling UTF-8 support in the regional settings, but I must say I didn't need to do that to get it working. Maybe it's necessary on older versions of Windows: https://superuser.com/a/1451686

Optionally we could also default Firefly-III to '' on Windows to use the system default settings (https://www.php.net/manual/en/function.setlocale.php) if nothing works. But I see someone submitted a pull request for it and it was rejected:  #3094

**Comment by danikuci1** [comment]
this is a new install. I should also explicitly state that there don't seem to be any usability issues, everything so far displays just fine.

OS is Windows Server 2019 Standard. checking default region settings shows that UTF-8 is not enabled:
![image](https://user-images.githubusercontent.com/5247319/97761773-d2a67000-1ac3-11eb-98d9-71d0fdb87144.png)

I've enabled it and rebooted, but am still getting the same repeating error in the log, and same banner at the top:
```
[2020-10-30 15:24:07] local.ERROR: Could not set locale. The following array doesnt work:  ["en-US.utf8","en-US.UTF-8"] 
[2020-10-30 15:24:14] local.ERROR: Could not set locale. The following array doesnt work:  ["en-US.utf8","en-US.UTF-8"] 
[2020-10-30 15:24:14] local.ERROR: Could not set locale. The following array doesnt work:  ["en-US.utf8","en-US.UTF-8"] 
[2020-10-30 15:24:15] local.ERROR: Could not set locale. The following array doesnt work:  ["en-US.utf8","en-US.UTF-8"] 
```


Is there a command or something that I could run to determine what locales the system says are available? Perhaps that could at least determine "en-us" vs. "en_us"

**Comment by danikuci1** [comment]
I was able to dig up [Get-WinSystemLocale](https://docs.microsoft.com/en-us/powershell/module/internationalcmdlets/get-winsystemlocale?view=winserver2012-ps) which seems to suggest that the correct locale is "en-US" and is available my system. This would seem to suggest this is a bug with Firefly III, would it not?
```
PS C:\Windows\system32> Get-WinSystemLocale

LCID             Name             DisplayName
----             ----             -----------
1033             en-US            English (United States)
```
As a work around, how can I suppress the error banner that appears at the top of every page?

**Comment by sephrat** [comment]
Mmmh, if I run this command it only tells me about fr-FR although I have the other locales installed and working with Firefly-III. Not sure this command is really helpful here.
My knowledge on Windows ends here, sorry. I won't be of better help than Google.

If you don't mind fiddling with the code, you could try to force the locale to a different value:
https://github.com/firefly-iii/firefly-iii/blob/e6c0bbf082c64ed7a2794930baf0861e1199d1fd/app/Support/Steam.php#L601
Replace `return $locale;` with something like `return 'en-US;'` and see how it goes.
Unless I'm mistaken, the default configuration for Firefly III uses the following value:
https://github.com/firefly-iii/firefly-iii/blob/e6c0bbf082c64ed7a2794930baf0861e1199d1fd/resources/lang/en_US/config.php#L27 
... although it replaces en_US with en-US if your server is Windows.

Hope this helps.

**Comment by danikuci1** [comment]
The cmdlet I was pointing out returns the current locale, not all that are available. My point was to prove that "en-US" was on my system, and that the problem lies somewhere in the codebase not picking it up properly.

Instead, I found this: https://github.com/firefly-iii/firefly-iii/blob/2c69a2eda2bfda39bde1eafd2d314a26f8d0d928/resources/views/v1/partials/flashes.twig and simply deleted the if statement for `invalidMonetaryLocale`. So I have a workaround, but i still think this is a bug with how Firefly III retrieves or interprets info about the system it is on.

I'd like to do some more digging though, since your installation is working. Could you provide me some info about your config?
* Could you share your OS build info? `systeminfo | findstr /C:"OS"`
* If you set your user locale to something else, are you able to get this error banner to pop-up?
* what locale and language settings are in your `.env` file? do they match your user settings?

TIA

**Comment by sephrat** [comment]
> Could you share your OS build info? `systeminfo | findstr /C:"OS"`

I'm running Microsoft Windows 10 Professionnel 10.0.19041 N/A version 19041.

>  If you set your user locale to something else, are you able to get this error banner to pop-up?

Good question, I can't even break my setup. I tried lots of obscure locale like Polish, Russian, Chinese, and they all work fine. So it's definitely not related to the regional settings I pointed out in a previous message.

> what locale and language settings are in your `.env` file? do they match your user settings?

The settings in `.env` are only there to set default values for new accounts. Your user preferences are the only settings that matter. 


If it's of any interest, I run Firefly-III with XAMPP.
